fairscale/nn/moe/moe_layer.py

```diff
@@ -79,15 +80,19 @@ def forward(self, *input: Tensor, **kwargs: Any) -> Tensor:
         assert input[0].shape[0] % len(self.experts) == 0, "num tokens must be order of number of local experts"
 
         # Implement Algorithm 2 from GShard paper.
-        shape = input[0].shape
+        d_model = input[0].shape[2]
         # Reshape into S tokens per group.
-        reshaped_input = input[0].reshape(-1, shape[2])
+        reshaped_input = input[0].reshape(-1, d_model)
         self.l_aux, combine_weights, dispatching_mask = self.gate(reshaped_input)
         dispatched_input = self.all_to_all_dispatch(dispatching_mask, reshaped_input)
-        chunks = dispatched_input.chunk(len(self.experts), dim=0)
+        # Re-shape after all-to-all: ecm -> gecm
+        dispatched_input = dispatched_input.reshape(self.world_size, self.num_local_experts, -1, d_model)
+        chunks = dispatched_input.chunk(len(self.experts), dim=1)
```

Comment:
Shouldn't this be `dispatched_input.chunk(len(self.num_local_experts), dim=1)` ?

Reply:
Fixed. The earlier code was also correct since:

len(self.experts) == self.num_local_experts

Reply:
ah, you are right! For some reason, I thought this was the global number of experts.